### ISSUE TITLE
Interaction between lock() and processors' store buffers

### ISSUE THREAD

**Issue by AndrewZ1989** [dotnet/coreclr#20636]
Hello everyone,

Let's suppose I have the following code:

    1   private readonly object lockObj= new object();
    2
    3   public bool IsInitialized { get; private set; }
    4
    5   public void DoMethod()
    6    {
    7       if( IsInitialized  ) return;
    8
    9        lock( lockObj )
    10       {
    11           if( IsInitialized  ) return;
    12
    13           //Do some stuff here
    14
    15           IsInitialized = true;
    16        }
    17    }

Suppose the Thread1 is inside the lock section on the line 13 and the Thread2 is waiting on the line 9. 

Is it guaranteed that after the Thread1 will leave the lock section the Thread2 will read 'true' on the line 11?
In other words, does Monitor.Enter() causes processors' store buffers flushing?

P.s.: personally I'd use volatile field in this case, but one person told me that the code above is also correct.

**Comment by mikedn** [comment]
> Is it guaranteed that after the Thread1 will leave the lock section the Thread2 will read 'true' on the line 11?

Yes. The docs do not say this explicitly but locks (pretty much any kind of sane lock, not only .NET's monitor) have, unsurprisingly, acquire/release semantic.
* Acquiring a lock implies that load operations cannot be moved before the acquire operation. In other words, the load at line 11 cannot ever be performed before the lock is acquired. 
* Releasing a lock implies that store operations cannot be moved after the release operation. That is, the store at line 15 cannot ever be performed after the lock is released.

Locks would be pretty much useless if the above weren't true. What happens inside the lock must stay inside the lock.

> In other words, does Monitor.Enter() causes processors' store buffers flushing?

Sort of. It's better not to drag low level details such as store buffers into this. What matters is only the order of the operations and that's usually described in terms of acquire/release semantics or happens-before relations. Intel processors (and probably others) do have this notion of store buffer and it's something that can impact concurrency. But maintaining load/store is something that the cache coherency protocol does and it's a bit more complicated than just "flushing buffers".

**Comment by jkotas** [comment]
The pattern you have described is called [Double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking).

The pattern you have described is non-portable. It happens to work in the current .NET x86/x64, but not on other platforms. The problem is not at lines 9 and 13. The problem is at lines 7 and 13: `if (IsInitialized) return;`. When you return on line 7, it is not guarantee that the actions done from "Do some stuff here" are visible from other threads, e.g. if "Do some stuff" initializes a static variable, you may still see the old uninitialized value once you return on line 7.

There are [many articles](https://www.bing.com/search?q=double%20checked%20locking%20in%20C%23) out there that discuss this issue. I think http://csharpindepth.com/Articles/General/Singleton.aspx is a pretty good one. I agree with the conclusion of the article that suggests to use static field initializers.
